### PR TITLE
[TritonGEN]: Add 2d block prefetch operator

### PR DESCRIPTION
**Pull request by etiotto**

### CODE REVIEW

File: test/TritonGEN/tritongen-to-llvm.mlir [hunk 1]
```diff
@@ -173,3 +173,12 @@ llvm.func @triton_gen.2Dblockstore(%ptr : !llvm.ptr, %base_width : i32, %base_he
   triton_gen.2Dblockstore %ptr, %base_width, %base_height, %base_pitch, %x, %y, %stored_val {elem_size_in_bits=32, tile_width=8, tile_height=8, v_blocks=1, transpose=false, vnni_transform=false} : (!llvm.ptr, i32, i32, i32, i32, i32, vector<8xf32>)
   llvm.return
 }
+
+// -----
+
+llvm.func @triton_gen.2Dblockprefetch(%ptr : !llvm.ptr, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
+  // CHECK:     llvm.func @triton_gen.2Dblockprefetch(%arg0: !llvm.ptr, %arg1: i32, %arg2: i32, %arg3: i32, %arg4: i32, %arg5: i32) {
+  // CHECK-NEXT:  triton_gen.2Dblockprefetch %arg0, %arg1, %arg2, %arg3, %arg4, %arg5 {cache_control = L1UC_L3UC} : (!llvm.ptr, i32, i32, i32, i32, i32)
```
Comment (whitneywhtsang) [inline test/TritonGEN/tritongen-to-llvm.mlir:181]: It is not lowered.

**Review by whitneywhtsang — CHANGES_REQUESTED** [review]

File: test/TritonGEN/tritongen-to-llvm.mlir (hunk 1 above)
Comment (etiotto) [reply test/TritonGEN/tritongen-to-llvm.mlir]: done

File: lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 2]
```diff
@@ -101,85 +176,28 @@ LogicalResult TritonGEN::MatrixDPASOp::verify() {
 // gen.2Dblockload
 //===----------------------------------------------------------------------===//
 
-static std::optional<int> getConstantInt(Value v) {
```
Comment (etiotto) [inline lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:104]: Just moved these utility functions at the start of the file

File: lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 3]
```diff
@@ -101,85 +176,28 @@ LogicalResult TritonGEN::MatrixDPASOp::verify() {
...
 }
 
 //===----------------------------------------------------------------------===//
-// gen.2Dblockstore
+// gen.2Dblockprefetch
 //===----------------------------------------------------------------------===//
 
-LogicalResult TritonGEN::Matrix2DBlockStoreOp::verify() {
-  return verifyInput(*this);
+LogicalResult TritonGEN::Matrix2DBlockPrefetchOp::verify() {
```
Comment (etiotto) [inline lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:195]: Need to add more verification checks here (and expand the lit tests).

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 4]
```diff
@@ -321,9 +319,67 @@ createGenISA2DBlockWrite(TritonGEN::Matrix2DBlockStoreOp op,
                           y,       elemSize,     tileWidth,     tileHeight,
                           vBlocks, useTranspose, vnniTransform, cache,
                           storeVal};
-  auto callOp = rewriter.create<LLVM::CallOp>(loc, funcOp, args);
+  return rewriter.create<LLVM::CallOp>(loc, funcOp, args);
+}
 
-  return callOp;
+static LLVM::CallOp
+createGenISA2DBlockPrefetch(TritonGEN::Matrix2DBlockPrefetchOp op,
```
Comment (etiotto) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:326]: Note that currently I have lowered the operator to a GenISA call. We will need to update the codegen to use OpenCL functions once they are available...

File: lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 5]
```diff
@@ -8,11 +8,86 @@
...
+  if (pitch && width && *pitch < *width)
+    return op->emitOpError(
+        "4th operand (base pitch) should be >= 2nd operand (base width)");
+
+  uint32_t TileWidth = op.getTileWidth();
+  uint32_t TileHeight = op.getTileHeight();
+  switch (op.getElemSizeInBits()) {
+  case 32:
+    if (TileWidth != 8)
+      return op->emitOpError("tile_width for 32 bit elements should be equal "
```
Comment (chengjunlu) [inline lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:63]: Can we use a more flexible checking? The 2D block load/store is for general purpose. Just protect the user input fits the HW's limitation is simple.

File: lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 6]
```diff
@@ -8,11 +8,86 @@
...
+#include "llvm/ADT/STLExtras.h"
 
 using namespace mlir;
 using namespace mlir::triton;
 
+//===----------------------------------------------------------------------===//
+// Utility functions
+//===----------------------------------------------------------------------===//
+
+static std::optional<int> getConstantInt(Value v) {
```
Comment (Dewei-Wang-sh) [inline lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:20]: mlir has "std::optional<int64_t> getConstantIntValue", we can use that alternatively.

**Review by Dewei-Wang-sh — APPROVED** [review]
lgtm

File: include/triton/Dialect/TritonGEN/IR/TritonGENOps.td [hunk 7]
```diff
@@ -315,7 +315,7 @@ def TritonGEN_Matrix2DBlockStoreOp : TritonGEN_Op<"2Dblockstore">,
     $ptr - the base address of the memory array
     $base_width, $base_height, $base_pitch - the shape of the memory array
     $x, $y, $tile_width, $tile_height - the starting offsets and shape of the submatrix to load
-    $elem_size_in_bits - 32 for f32, bf32; 16 for f16, int16, bf16; 8 for int8, int4, int2 and etc
+    $elem_size_in_bits - 32 for f32, bf32; 16 for f16, int16, bf16; 8 for int8 etc...
```
Comment (victor-eds) [inline include/triton/Dialect/TritonGEN/IR/TritonGENOps.td:318]:
```suggestion
    $elem_size_in_bits - 32 for f32, bf32; 16 for f16, int16, bf16; 8 for int8, etc...
```

File: lib/Dialect/TritonGEN/IR/TritonGENOps.cpp (hunk 6 above)
Comment (victor-eds) [reply lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:20]: [`constant_int_value_binder`](https://github.com/llvm/llvm-project/blob/577e0ef94fb0b4ba9f97a6f58a1961f7ba247d21/mlir/include/mlir/IR/Matchers.h#L176) and [`constant_int_predicate_matcher`](https://github.com/llvm/llvm-project/blob/577e0ef94fb0b4ba9f97a6f58a1961f7ba247d21/mlir/include/mlir/IR/Matchers.h#L208) might also be a good fit for this particular use case, but we can use what @Dewei-Wang-sh mentions.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 8]
```diff
@@ -321,9 +319,67 @@ createGenISA2DBlockWrite(TritonGEN::Matrix2DBlockStoreOp op,
...
+  Location loc = op->getLoc();
+
+  Value ptr = op.getPtr();
+  Value baseWidth = op.getBaseWidth();
+  Value baseHeight = op.getBaseHeight();
+  Value basePitch = op.getBasePitch();
+  Value x = op.getX();
+  Value y = op.getY();
+
+  std::string funcName = "llvm.genx.GenISA.LSC2DBlockPrefetch.isVoid";
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:339]:
```suggestion
  constexpr StringLiteral funcName = "llvm.genx.GenISA.LSC2DBlockPrefetch.isVoid";
```

File: lib/Dialect/TritonGEN/IR/TritonGENOps.cpp (hunk 6 above)
Comment (etiotto) [reply lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:20]: Changed to use `getConstantIntValue`

File: include/triton/Dialect/TritonGEN/IR/TritonGENOps.td (hunk 7 above)
Comment (etiotto) [reply include/triton/Dialect/TritonGEN/IR/TritonGENOps.td:318]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 8 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:339]: done

File: lib/Dialect/TritonGEN/IR/TritonGENOps.cpp (hunk 5 above)
Comment (etiotto) [reply lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:63]: Sure, this PR is for adding the prefetch operation. Added an issue to track this request.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**